### PR TITLE
sql: division by zero + TypeCheckExpr

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

**Comment by tamird** [comment]
@petermattis I'm not sure how to reconcile this with your comment re TypeCheckExpr in analyze.go. That comment is sitting on a function that specifically handles tuples, which are decidedly poorly handled in this implementation.

Pointers are appreciated.

File: sql/parser/eval.go [hunk 1]
```diff
@@ -581,6 +605,380 @@ func init() {
 	cmpOps[cmpArgs{In, tupleType, tupleType}] = evalTupleIN
 }
 
+// TypeCheckExpr type-cheks an SQL expression. Expression checking is a
```
Comment (petermattis) [inline sql/parser/eval.go]: `/cheks/checks/g`

File: sql/parser/eval.go [hunk 2]
```diff
@@ -581,6 +605,380 @@ func init() {
...
+	case *QualifiedName:
+		return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+
+	case Tuple:
+		// TODO(tamird): handling of tuples leaves much to be desired. This
+		// is largely a consequence of go's type system not being
+		// expressive enough to encode the type of a heterogenous tuple
+		// (dBool, dString) as distinct from the tuple (dBool). We'll have
+		// to write our own type system here to do this properly.
+		return tupleType, nil
```
Comment (petermattis) [inline sql/parser/eval.go]: I think that using `reflect.Type` as the return value from `TypeCheckExpr` is the root of the problem. I think it would be better for `TypeCheckExpr` to return a `Datum`, the type of which is the type of the expression. Then you could return the tuple itself here after checking that all of the tuple elements have compatible types (i.e. are either `dNull` or the same type).

File: sql/parser/eval.go [hunk 3]
```diff
@@ -581,6 +605,380 @@ func init() {
...
+	case GE:
+		// GE(left, right) is implemented as LE(right, left)
+		op = LE
+		leftType, rightType = rightType, leftType
+	case NotIn:
+		// NotIn(left, right) is implemented as !IN(left, right)
+		op = In
+	}
+
+	// TODO(pmattis): Memoize the cmpOps lookup as we've done for unaryOps and
```
Comment (petermattis) [inline sql/parser/eval.go]: Delete this comment. I don't think it applies here.

File: sql/parser/eval.go [hunk 4]
```diff
@@ -581,6 +605,380 @@ func init() {
 	cmpOps[cmpArgs{In, tupleType, tupleType}] = evalTupleIN
 }
 
+// TypeCheckExpr type-cheks an SQL expression. Expression checking is a
+// mostly straightforward walk over the parse tree.
+// TODO(tamird): tuple handling is currently crappy.
+func TypeCheckExpr(expr Expr) (reflect.Type, error) {
```
Comment (petermattis) [inline sql/parser/eval.go]: `eval.go` has grown to be a bloated beast. Let's put this in a new `type_check.go` file. I think it is also time to move all of the datum stuff into a `datum.go` file, but let's hold off on doing that until Vivek's date/timestamp change is in.

File: sql/parser/eval.go [hunk 5]
```diff
@@ -771,40 +1166,26 @@ func evalRangeCond(expr *RangeCond) (Datum, error) {
...
-		return DNull, err
-	}
-
-	limits := [2]struct {
-		op   ComparisonOp
-		expr Expr
-	}{
-		{GE, expr.From},
-		{LE, expr.To},
+	var newExpr Expr = &AndExpr{
```
Comment (petermattis) [inline sql/parser/eval.go]: Why this change? It doesn't seem related to this type-checking. More to the point, this code shouldn't even be necessary anymore now that `NormalizeExpr` does exactly this normalization.

File: sql/parser/normalize.go [hunk 6]
```diff
@@ -281,6 +281,8 @@ func (v *normalizeVisitor) normalizeComparisonExpr(n *ComparisonExpr) (Visitor,
 				} else {
 					left.Operator = Mult
 				}
+				// Bust the function cache now that we've changed the operator.
```
Comment (petermattis) [inline sql/parser/normalize.go]: `s/Bust/Clear/g`

Comment (petermattis) [inline sql/parser/normalize.go]: Can you add a test that demonstrates a bug caused by this?

File: sql/scan.go [hunk 7]
```diff
@@ -350,6 +353,12 @@ func (n *scanNode) initTargets(targets parser.SelectExprs) error {
 			return n.err
 		}
 	}
+	for _, render := range n.render {
+		if _, n.err = parser.TypeCheckExpr(render); n.err != nil {
```
Comment (petermattis) [inline sql/scan.go]: Any reason not to call `TypeCheckExpr` before calling `addRender`? Or perhaps, call `TypeCheckExpr` in `addRender`.

File: sql/select.go [hunk 8]
```diff
@@ -49,6 +49,7 @@ func (p *planner) Select(n *parser.Select) (planNode, error) {
 	if err := scan.initTargets(n.Exprs); err != nil {
 		return nil, err
 	}
+
```
Comment (petermattis) [inline sql/select.go]: This function grew organically which resulted in the lack of spaces, but placing a single space here seems spurious. Perhaps spaces between each of the `scan.init` sections and between group-by and order-by processing.

**Comment by petermattis** [comment]
Lots of overlap between `TypeCheckExpr` and `EvalExpr`. I'm fine with copying code and making small changes to it, but its hard to see where the changes are. Not sure if there is some git magic to make this more obvious. Perhaps when `TypeCheckExpr` is moved to `type_check.go`.

File: sql/scan.go (hunk 7 above)
Comment (tamird) [inline sql/scan.go]: > Any reason not to call `TypeCheckExpr` before calling `addRender`?

addRender substitutes qualified names, which is required by `TypeCheckExpr`.

> Or perhaps, call TypeCheckExpr in addRender.

This is possible, but just a bit ugly since there's an if-else and the call would be required in both branches.

File: sql/parser/normalize.go (hunk 6 above)
Comment (tamird) [inline sql/parser/normalize.go]: one of the order_by tests caught this

File: sql/parser/eval.go (hunk 5 above)
Comment (tamird) [inline sql/parser/eval.go]: It was a simplification that I stumbled upon when I was writing the typeCheck equivalent of this function. Given that NormalizeExpr does this, I'll just remove it altogether.

File: sql/parser/eval.go (hunk 3 above)
Comment (tamird) [inline sql/parser/eval.go]: I think it does - this is where the caching will happen.

File: sql/scan.go (hunk 7 above)
Comment (tamird) [inline sql/scan.go]: nevermind, found a good place for it in `addRender`.

File: sql/select.go (hunk 8 above)
Comment (tamird) [inline sql/select.go]: totally spurious, removed

File: sql/parser/eval.go (hunk 1 above)
Comment (tamird) [inline sql/parser/eval.go]: fixed

File: sql/parser/eval_test.go [hunk 9]
```diff
@@ -238,3 +239,42 @@ func TestEvalExprError(t *testing.T) {
 		}
 	}
 }
+
+func TestTypeCheckExprError(t *testing.T) {
```
Comment (petermattis) [inline sql/parser/eval_test.go]: This belongs in `type_check_test.go`.

File: sql/parser/type_check.go [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+	"reflect"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// TypeCheckExpr type-checks an SQL expression. Expression checking is
+// a mostly straightforward walk over the parse tree.
+// TODO(tamird): tuple handling is currently crappy.
+func TypeCheckExpr(expr Expr) (reflect.Type, error) {
```
Comment (petermattis) [inline sql/parser/type_check.go]: My comment about using `Datum` to hold type information instead of `reflect.Type` seems to have gotten lost. If you returned a `Datum` here then you would be able to distinguish between different types of tuples.

File: sql/parser/eval.go (hunk 3 above)
Comment (petermattis) [inline sql/parser/eval.go]: Ah, I see. You've moved all of the caching to the `typeCheck` routines. Ok.

File: sql/parser/type_check.go [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+	switch op {
+	case Like, NotLike:
+		return nil, util.Errorf("TODO(pmattis): unsupported comparison operator: %s", op)
+	}
+
+	return nil, fmt.Errorf("unsupported comparison operator: <%s> %s <%s>",
+		typeToValue[leftType].Type(), op, typeToValue[rightType].Type())
+}
+
+func typeCheckRangeCond(expr *RangeCond) (reflect.Type, error) {
```
Comment (petermattis) [inline sql/parser/type_check.go]: This method wouldn't be necessary if we called `NormalizeExpr` before `TypeCheckExpr`.

File: sql/scan.go [hunk 12]
```diff
@@ -350,6 +353,7 @@ func (n *scanNode) initTargets(targets parser.SelectExprs) error {
 			return n.err
 		}
 	}
+
```
Comment (petermattis) [inline sql/scan.go]: Spurious blank line introduced.

File: sql/scan.go [hunk 13]
```diff
@@ -447,8 +451,8 @@ func (n *scanNode) addRender(target parser.SelectExpr) error {
 	if resolved, n.err = n.resolveQNames(target.Expr); n.err != nil {
 		return n.err
 	}
-	// Evaluate the expression once to memoize operators and functions.
-	if _, n.err = parser.EvalExpr(resolved); n.err != nil {
+	// Type check the expression to memoize operators and functions.
```
Comment (petermattis) [inline sql/scan.go]: We should be calling `NormalizeExpr` here too. Not doing so was unintentional.

File: sql/parser/type_check.go [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+	expr.rtype = rightType
+	left := typeToValue[leftType]
+	right := typeToValue[rightType]
+
+	var ok bool
+	if expr.fn, ok = binOps[binArgs{expr.Operator, leftType, rightType}]; !ok {
+		return nil, fmt.Errorf("unsupported binary operator: <%s> %s <%s>",
+			left.Type(), expr.Operator, right.Type())
+	}
+	d, err := expr.fn(left, right)
```
Comment (petermattis) [inline sql/parser/type_check.go]: Rather the calling the function, perhaps we should expand `binOps` to hold both the function and the return value type. I think this would remove the need for the `dInt` hack.

File: sql/parser/type_check.go [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+		leftType, rightType = rightType, leftType
+	case NotIn:
+		// NotIn(left, right) is implemented as !IN(left, right)
+		op = In
+	}
+
+	// TODO(pmattis): Memoize the cmpOps lookup as we've done for unaryOps and
+	// binOps.
+	if f, ok := cmpOps[cmpArgs{op, leftType, rightType}]; ok {
+		return reflect.TypeOf(f).Out(0), nil
```
Comment (petermattis) [inline sql/parser/type_check.go]: Won't `reflect.TypeOf(f).Out(0)` always be `boolType`?

File: sql/parser/type_check.go [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+		return nil, fmt.Errorf("unsupported unary operator: %s <%s>",
+			expr.Operator, arg.Type())
+	}
+
+	d, err := expr.fn(arg)
+	if err != nil {
+		return nil, err
+	}
+
+	return reflect.TypeOf(d), nil
```
Comment (petermattis) [inline sql/parser/type_check.go]: For unary ops, the output type is currently always the same as the input type. But like `binOps` I'd prefer if we expanded `unaryOps` to also contain the type of the return value.

**Comment by petermattis** [comment]
A few more nits and one major issue with using `reflect.Type` vs `Datum`. 

I'm glad you tackled this issue, though we almost duplicated the work. #2228 is assigned to me and I was planning to tackle it yesterday but got bogged down in school/move work. Next time, perhaps ping the bug and see if the person it is assigned to has started on it.

**Comment by vivekmenezes** [comment]
This is a nice change.

**Comment by tamird** [comment]
@petermattis ah, sorry, I was coming at this from #1622.

File: sql/parser/eval_test.go (hunk 9 above)
Comment (tamird) [inline sql/parser/eval_test.go]: moved

File: sql/parser/type_check.go (hunk 11 above)
Comment (tamird) [inline sql/parser/type_check.go]: removed

File: sql/scan.go (hunk 12 above)
Comment (tamird) [inline sql/scan.go]: thanks, removed

File: sql/scan.go (hunk 13 above)
Comment (tamird) [inline sql/scan.go]: yeah, fixed this by introducing NormalizeAndTypeCheckExpr and calling that.

File: sql/parser/type_check.go (hunk 15 above)
Comment (tamird) [inline sql/parser/type_check.go]: yeah, but I felt this was nicer than duplicating that fact (which is why i changed the type of cmpOps to return DBool rather than Datum)

File: sql/parser/type_check.go (hunk 14 above)
Comment (tamird) [inline sql/parser/type_check.go]: Good call, done. I also had to add returnType to `builtin` as well to get the dInt hack all the way out.

Unfortunately not all functions have knowable types, so I still end up calling those functions that have a nil `returnType`

File: sql/parser/type_check.go (hunk 16 above)
Comment (tamird) [inline sql/parser/type_check.go]: done

**Comment by tamird** [comment]
addressed everything (i think) except for the reflect.Type vs datum issue, which I'll take a look at after lunch.

File: sql/parser/eval.go [hunk 17]
```diff
@@ -591,13 +726,6 @@ func init() {
 // the client.
 func EvalExpr(expr Expr) (Datum, error) {
 	switch t := expr.(type) {
-	case Row:
```
Comment (petermattis) [inline sql/parser/eval.go]: Add a `case` for `Row` that indicates that `// NormalizeExpr transforms this into Row`.

File: sql/parser/type_check.go (hunk 15 above)
Comment (petermattis) [inline sql/parser/type_check.go]: At least add a comment. It isn't at all obvious that `reflect.TypeOf(f).Out(0)` will always be `boolType`.

File: sql/parser/type_check.go [hunk 18]
```diff
@@ -0,0 +1,358 @@
...
+	case *QualifiedName:
+		return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+
+	case Tuple:
+		// TODO(tamird): handling of tuples leaves much to be desired. This
+		// is largely a consequence of go's type system not being
+		// expressive enough to encode the type of a heterogenous tuple
+		// (dBool, dString) as distinct from the tuple (dBool). We'll have
+		// to write our own type system here to do this properly.
+		return tupleType, nil
```
Comment (petermattis) [inline sql/parser/type_check.go]: You should at least loop over the tuple elements and verify they have the same type or the `nullType`.

File: sql/parser/eval.go [hunk 19]
```diff
@@ -581,6 +699,23 @@ func init() {
 	cmpOps[cmpArgs{In, tupleType, tupleType}] = evalTupleIN
 }
 
+// NormalizeAndTypeCheckExpr is a combination of NormalizeExpr and
```
Comment (petermattis) [inline sql/parser/eval.go]: Slightly strange that this function is in `eval.go` instead of either `normalize.go` or `type_check.go`.

File: sql/parser/eval.go (hunk 17 above)
Comment (tamird) [inline sql/parser/eval.go]: done

File: sql/parser/type_check.go (hunk 15 above)
Comment (tamird) [inline sql/parser/type_check.go]: done

File: sql/parser/type_check.go (hunk 18 above)
Comment (tamird) [inline sql/parser/type_check.go]: I'm doing the `Datum` thing, so that this will be proper.

**Comment by petermattis** [comment]
Let's leave the `reflect.Type` to `Datum` change to a follow-on PR.

LGTM, though you and @vivekmenezes will need to fight over who merges first. Addressing the merge conflicts could be irritating. I suggest you let Vivek merge first as you are the better git wizard.

**Comment by tamird** [comment]
@petermattis this is ready for a final look.

File: sql/parser/eval.go (hunk 19 above)
Comment (petermattis) [inline sql/parser/eval.go]: Ping? Not a big deal, just seems misplaced.

File: sql/parser/type_check.go [hunk 20]
```diff
@@ -0,0 +1,432 @@
...
+			}
+		} else if op == In && rType == tupleType {
+			if err := typeCheckTupleIN(dummyLeft, dummyRight); err != nil {
+				return nil, err
+			}
+		}
+
+		// reflect.TypeOf(f).Out(0) is always boolType, but no reason to
+		// duplicate that here.
+		return reflect.New(reflect.TypeOf(f).Out(0)).Elem().Interface().(Datum), nil
```
Comment (petermattis) [inline sql/parser/type_check.go]: I disagree with the comment now. `reflect.New(reflect.TypeOf(f).Out(0)).Elem().Interface()` is quite a mouthful in comparison to `DummyBool`.

File: sql/parser/type_check.go [hunk 21]
```diff
@@ -0,0 +1,432 @@
...
+		// NotIn(left, right) is implemented as !IN(left, right)
+		op = In
+	}
+
+	lType := reflect.TypeOf(dummyLeft)
+	rType := reflect.TypeOf(dummyRight)
+
+	// TODO(pmattis): Memoize the cmpOps lookup as we've done for unaryOps and
+	// binOps.
+	if f, ok := cmpOps[cmpArgs{op, lType, rType}]; ok {
```
Comment (petermattis) [inline sql/parser/type_check.go]: It looks like this will fail to find a comparison op if either argument is `DNull`. I'm not sure that is correct. `a = NULL` is ok from a type-checking perspective, but semantically always `NULL`.

Comment (tamird) [inline sql/parser/type_check.go]: Fixed + test.

File: sql/parser/type_check.go (hunk 20 above)
Comment (tamird) [inline sql/parser/type_check.go]: addressed by extracting a package-scoped var. it's a mouthful, but just in one place

File: sql/parser/eval.go (hunk 19 above)
Comment (tamird) [inline sql/parser/eval.go]: Moved to type_check.go

File: sql/parser/eval.go [hunk 22]
```diff
@@ -525,45 +634,47 @@ type cmpArgs struct {
 	rightType reflect.Type
 }
 
+var cmpOpResultType = reflect.New(reflect.TypeOf(cmpOps).Elem().Out(0)).Elem().Interface().(Datum)
```
Comment (petermattis) [inline sql/parser/eval.go]: All that reflect stuff so you can change the type of the return value for the `cmpOps` functions without adjusting this type. Doesn't seem worth it to me. I challenge anyone who is not well versed in `reflect` to understand this.

File: sql/parser/type_check_test.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+package parser
+
+import (
+	"regexp"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/testutils"
+)
+
+func TestTypeCheckExp(t *testing.T) {
```
Comment (petermattis) [inline sql/parser/type_check_test.go]: `s/TestTypeCheckExp/TestTypeCheckExpr/g`? Or perhaps you meant something else with the `Exp` suffix.

File: sql/testdata/order_by [hunk 24]
```diff
@@ -135,7 +135,7 @@ SELECT a FROM t ORDER BY (((a)))
 4
 5
 
-query error incompatible value type bool
+query error incompatible value types bool, int
 SELECT CASE a WHEN 1 THEN b ELSE c END as val FROM t ORDER BY val
 ----
 1
```
Comment (petermattis) [inline sql/testdata/order_by]: The query values shouldn't be needed when there is an error.

**Comment by petermattis** [comment]
LGTM

File: sql/parser/type_check_test.go (hunk 23 above)
Comment (tamird) [inline sql/parser/type_check_test.go]: just a mistake, fixed.

File: sql/parser/eval.go (hunk 22 above)
Comment (tamird) [inline sql/parser/eval.go]: More than just saving the change, I think of this as a static assertion

Comment (petermattis) [inline sql/parser/eval.go]: If you want a static assertion, change the `.(Datum)` to `.(DBool)`.

Comment (tamird) [inline sql/parser/eval.go]: good call.

File: sql/testdata/order_by (hunk 24 above)
Comment (tamird) [inline sql/testdata/order_by]: fixed. squashing and merging
